save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/utils/KotlinUtils.kt

```diff
@@ -18,15 +19,41 @@ typealias StringList = List<String>
  * @param action action that should be invoked
  * @return [T] if the result was fetched in [times] attempts, null otherwise
  */
+@Suppress("TooGenericExceptionCaught")
 suspend fun <T : Any> retry(
     times: Int,
     delayMillis: Long = 10_000L,
-    action: () -> T?,
-): T? = action() ?: run {
-    if (times > 0) {
-        delay(delayMillis)
-        retry(times - 1, delayMillis, action)
-    } else {
-        null
+    action: suspend (Int) -> T?,
+): Pair<T?, List<Throwable>> {
+    val caughtExceptions: MutableList<Throwable> = mutableListOf()
+    times.downTo(0).map { iteration ->
+        try {
+            delay(delayMillis)
```

Comment:
you make delay before the first attempt -- is it required?

Reply:
No, it is not
Will fix it

Reply:
Moved `delay` to `finally` block